### PR TITLE
[dockerng] Add support for volume management in dockerng

### LINKED ISSUE #28990: Dockerng volume
https://blog.docker.com/2015/11/docker-1-9-production-ready-swarm-multi-host-networking/

Docker 1.9 has a new volume management Api:

    Usage:  docker volume [OPTIONS] [COMMAND]

    Manage Docker volumes

    Commands:
      create                 Create a volume
      inspect                Return low-level information on a volume
      ls                     List volumes
      rm                     Remove a volume

    Run 'docker volume COMMAND --help' for more information on a command

      --help=false       Print usage

This is very convenient and in my opinion much needed for orchestration. Saltstack has been great at docker orchestration thus Dockerng should be extended to manage volume as well.

### PR DESCRIPTION
**Pull request by ticosax**
fixes  #28990

### CODE REVIEW

**Comment by cachedout** [comment]
Hi @ticosax Looks like a few Python 3 incompatability issues. Please take a look at the lint report and let me know what you think. Thanks!

**Comment by ticosax** [comment]
@cachedout thanks for the notification. 
I pushed the fix

**Comment by cachedout** [comment]
@ticosax Thank you for your quick response. Once the tests finish running I will have another look.

File: salt/modules/dockerng.py [hunk 1]
```diff
@@ -4419,6 +4424,99 @@ def disconnect_container_from_network(container, network_id):
...
 
+# Volume Management
+
+
+@_api_version(1.21)
+@_client_version('1.5.0')
+def volumes(filters=None):
+    '''
+    List existing volumes
+
```
Comment (jfindlay) [inline salt/modules/dockerng.py]: @ticosax, `.. versionadded:: 2015.8.4`.

File: salt/modules/dockerng.py [hunk 2]
```diff
@@ -4419,6 +4424,99 @@ def disconnect_container_from_network(container, network_id):
...
+    # Only non-error return case is a True return, so just return the response
+    return response
+
+
+@_api_version(1.21)
+@_client_version('1.5.0')
+def create_volume(name, driver=None, driver_opts=None):
+    '''
+    Create a new volume
+
```
Comment (jfindlay) [inline salt/modules/dockerng.py]: @ticosax, `.. versionadded:: 2015.8.4`.

File: salt/modules/dockerng.py [hunk 3]
```diff
@@ -4419,6 +4424,99 @@ def disconnect_container_from_network(container, network_id):
...
+    # Only non-error return case is a True return, so just return the response
+    return response
+
+
+@_api_version(1.21)
+@_client_version('1.5.0')
+def remove_volume(name):
+    '''
+    Remove a volume
+
```
Comment (jfindlay) [inline salt/modules/dockerng.py]: @ticosax, `.. versionadded:: 2015.8.4`.

File: salt/modules/dockerng.py [hunk 4]
```diff
@@ -4419,6 +4424,99 @@ def disconnect_container_from_network(container, network_id):
...
+    # Only non-error return case is a True return, so just return the response
+    return response
+
+
+@_api_version(1.21)
+@_client_version('1.5.0')
+def inspect_volume(name):
+    '''
+    Inspect Network
+
```
Comment (jfindlay) [inline salt/modules/dockerng.py]: @ticosax, `.. versionadded:: 2015.8.4`.

File: salt/states/dockerng.py [hunk 5]
```diff
@@ -2081,6 +2081,128 @@ def network_absent(name, driver=None):
     return ret
 
 
+def volume_present(name, driver=None, driver_opts=None):
+    '''
+    Ensure that a volume is present.
+
```
Comment (jfindlay) [inline salt/states/dockerng.py]: @ticosax, `.. versionadded:: 2015.8.4`.

File: salt/states/dockerng.py [hunk 6]
```diff
@@ -2081,6 +2081,128 @@ def network_absent(name, driver=None):
...
+    ret['result'] = True
+    ret['comment'] = 'Volume \'{0}\' already exists.'.format(name)
+    return ret
+
+
+def volume_absent(name, driver=None):
+    '''
+    Ensure that a volume is absent.
+
```
Comment (jfindlay) [inline salt/states/dockerng.py]: @ticosax, `.. versionadded:: 2015.8.4`.

**Comment by jfindlay** [comment]
@ticosax, I have added some comments, but otherwise, this looks really awesome.

**Comment by ticosax** [comment]
@jfindlay Thanks, I amended the commits.

**Comment by cachedout** [comment]
Looks great. Thanks @ticosax !

**Comment by ticosax** [comment]
@cachedout thanks, #28990 can be closed.

**Pull request merged**